packages/horizon/contracts/payments/collectors/RecurringCollector.sol

```diff
@@ -0,0 +1,481 @@
+     * @param revokeSignerThawingPeriod The duration (in seconds) in which a signer is thawing before they can be revoked.
+     */
+    constructor(
+        string memory eip712Name,
+        string memory eip712Version,
+        address controller,
+        uint256 revokeSignerThawingPeriod
+    ) EIP712(eip712Name, eip712Version) GraphDirectory(controller) Authorizable(revokeSignerThawingPeriod) {}
+
+    /**
+     * @notice Initiate a payment collection through the payments protocol.
+     * See {IGraphPayments.collect}.
+     * @dev Caller must be the data service the RCA was issued to.
+     */
+    function collect(IGraphPayments.PaymentTypes paymentType, bytes calldata data) external returns (uint256) {
+        require(
+            paymentType == IGraphPayments.PaymentTypes.IndexingFee,
+            RecurringCollectorInvalidPaymentType(paymentType)
+        );
+        try this.decodeCollectData(data) returns (CollectParams memory collectParams) {
+            return _collect(collectParams);
+        } catch {
+            revert RecurringCollectorInvalidCollectData(data);
+        }
+    }
+
+    /**
+     * @notice Accept an indexing agreement.
+     * See {IRecurringCollector.accept}.
+     * @dev Caller must be the data service the RCA was issued to.
+     */
+    function accept(SignedRCA calldata signedRCA) external {
+        require(signedRCA.rca.agreementId != bytes16(0), RecurringCollectorAgreementIdZero());
+        require(
+            msg.sender == signedRCA.rca.dataService,
+            RecurringCollectorUnauthorizedCaller(msg.sender, signedRCA.rca.dataService)
+        );
+        require(
+            signedRCA.rca.deadline >= block.timestamp,
+            RecurringCollectorAgreementDeadlineElapsed(signedRCA.rca.deadline)
+        );
+
+        // check that the voucher is signed by the payer (or proxy)
+        _requireAuthorizedRCASigner(signedRCA);
+
+        AgreementData storage agreement = _getAgreementStorage(signedRCA.rca.agreementId);
+        // check that the agreement is not already accepted
+        require(
+            agreement.state == AgreementState.NotAccepted,
+            RecurringCollectorAgreementIncorrectState(signedRCA.rca.agreementId, agreement.state)
+        );
+
+        // accept the agreement
+        agreement.acceptedAt = uint64(block.timestamp);
+        agreement.state = AgreementState.Accepted;
+        agreement.dataService = signedRCA.rca.dataService;
+        agreement.payer = signedRCA.rca.payer;
+        agreement.serviceProvider = signedRCA.rca.serviceProvider;
+        agreement.endsAt = signedRCA.rca.endsAt;
+        agreement.maxInitialTokens = signedRCA.rca.maxInitialTokens;
+        agreement.maxOngoingTokensPerSecond = signedRCA.rca.maxOngoingTokensPerSecond;
+        agreement.minSecondsPerCollection = signedRCA.rca.minSecondsPerCollection;
+        agreement.maxSecondsPerCollection = signedRCA.rca.maxSecondsPerCollection;
+        _requireValidAgreement(agreement);
```

Comment:
nit: usually you'd check the inputs before doing any storage ops, consider changing the order of operations here.